execution_engine/src/system/auction.rs

```diff
@@ -528,93 +528,93 @@ pub trait Auction:
+                U512::from(*recipient.delegation_rate()),
+                U512::from(DELEGATION_RATE_DENOMINATOR),
+            );
+            let reward_multiplier: Ratio<U512> = Ratio::new(delegator_total_stake, total_stake);
+            let delegator_reward: Ratio<U512> = total_reward
+                .checked_mul(&reward_multiplier)
+                .ok_or(Error::ArithmeticOverflow)?;
+            let commission: Ratio<U512> = delegator_reward
+                .checked_mul(&commission_rate)
+                .ok_or(Error::ArithmeticOverflow)?;
+            delegator_reward
+                .checked_sub(&commission)
+                .ok_or(Error::ArithmeticOverflow)?
+        };
 
-            let delegator_rewards =
-                recipient
-                    .delegator_stake()
-                    .iter()
-                    .map(|(delegator_key, delegator_stake)| {
-                        let reward_multiplier = Ratio::new(*delegator_stake, delegator_total_stake);
-                        let reward = delegators_part * reward_multiplier;
-                        (delegator_key.clone(), reward)
-                    });
-            let delegator_payouts = detail::reinvest_delegator_rewards(
-                self,
-                seigniorage_allocations,
-                public_key.clone(),
-                delegator_rewards,
-            )?;
-            let total_delegator_payout: U512 = delegator_payouts
+        let delegator_rewards =
+            recipient
+                .delegator_stake()
                 .iter()
-                .map(|(_delegator_hash, amount, _bonding_purse)| *amount)
-                .sum();
-
-            let validators_part: Ratio<U512> = total_reward - Ratio::from(total_delegator_payout);
-            let validator_reward = validators_part.to_integer();
-            let validator_bonding_purse = detail::reinvest_validator_reward(
-                self,
-                seigniorage_allocations,
-                public_key.clone(),
-                validator_reward,
-            )?;
-
-            self.mint_into_existing_purse(validator_reward, validator_bonding_purse)
-                .map_err(Error::from)?;
+                .map(|(delegator_key, delegator_stake)| {
+                    let reward_multiplier = Ratio::new(*delegator_stake, delegator_total_stake);
+                    let reward = delegators_part * reward_multiplier;
+                    (delegator_key.clone(), reward)
+                });
+        let delegator_payouts = detail::reinvest_delegator_rewards(
+            self,
+            seigniorage_allocations,
+            proposer.clone(),
+            delegator_rewards,
+        )?;
+        let total_delegator_payout: U512 = delegator_payouts
+            .iter()
+            .map(|(_delegator_hash, amount, _bonding_purse)| *amount)
+            .sum();
+
+        let validators_part: Ratio<U512> = total_reward - Ratio::from(total_delegator_payout);
```

Comment:
Is this guaranteed to not underflow, because `to_integer` rounds towards zero, so `total_delegator_payout` will always be less than `total_reward`?